sys/kern/imgact_elf.c

```diff
@@ -511,25 +509,61 @@ __elfN(check_header)(const Elf_Ehdr *hdr)
 	return (0);
 }
 
+#if __has_feature(capabilities)
+/*
+ * Create a reservation for the image activator to map the executable
+ * image into.
+ */
 static int
-__elfN(copyout)(vm_map_t map, caddr_t kaddr, vm_offset_t uaddr, size_t sz)
+__elfN(build_imgact_capability)(struct image_params *imgp,
+    void * __capability *imgact_cap, const Elf_Ehdr *hdr, const Elf_Phdr *phdr,
+    u_long rbase)
 {
-	int error;
+	u_long perm = CHERI_PERM_STORE | CHERI_PERM_GLOBAL |
+	    CHERI_PERM_STORE_CAP;
+	vm_offset_t start = (vm_offset_t)-1;
+	vm_offset_t end = 0;
+	vm_offset_t seg_addr;
+	vm_size_t seg_size;
+	int i, result;
+	vm_ptr_t reservation;
+        void * __capability reservation_cap;
+	vm_map_t map;
+
+	map = &imgp->proc->p_vmspace->vm_map;
+
+	for (i = 0; i < hdr->e_phnum; i++) {
+		if (phdr[i].p_type != PT_LOAD || phdr[i].p_memsz == 0)
+			continue;
+
+		seg_addr = trunc_page(phdr[i].p_vaddr + rbase);
+		seg_size = round_page(phdr[i].p_memsz + phdr[i].p_vaddr +
+		    rbase - seg_addr);
+		start = MIN(start, seg_addr);
+		end = MAX(end, seg_addr + seg_size);
+	}
+
+	reservation = start;
+	result = vm_map_reservation_create(map, &reservation, end - start,
```

Comment:
Same `VM_PROT_RW` question.  It seems odd that the reservation would be `RW` but permit `RWX`?  It seems like either the reservation should enforce that `maxprot` is a subset of the reservation, or it should just not have permissions at all.

Reply:
Same as `imgact_aout`.